lib/view_component/collection.rb

```diff
@@ -26,5 +28,24 @@ def collection_variable(object)
         raise ArgumentError.new("The value of the argument isn't a valid collection. Make sure it responds to to_ary: #{object.inspect}")
       end
     end
+
+    def collection_parameter_name
+      @component.collection_parameter_name
+    end
+
+    def collection_counter_parameter_name
+      "#{collection_parameter_name}_counter".to_sym
+    end
+
+    def component_options(item, iterator)
+      item_options = { collection_parameter_name => item }
+      item_options[collection_counter_parameter_name] = iterator.index + 1 if counter_argument_present?
+
+      @options.merge(item_options)
+    end
+
+    def counter_argument_present?
```

Comment:
Can we set this value at compile time instead of at run time?

Reply:
I am not certain what you mean with this comment @joelhawksley :/

Reply:
Since the method signature of a component is static, we could set the result of this method as part of `compile` (in `Base`) so that we don't have to do this work at runtime.

Reply:
I hope this is what you mean @joelhawksley 1c168aa737bfbdcd357777e7bce9ebefbb2ef3c8

Reply:
I think I need to move it into the `compile` method now that I'm reading the code a bit more?

Reply:
Here is attempt 2 7c0676b87e063fe214a2605cb41bc708ea7778ea